chainer/optimizers/eve.py

```diff
@@ -0,0 +1,182 @@
+from __future__ import division
+import math
+
+import numpy
+
+from chainer import optimizer
+from chainer.optimizers import adam
+
+
+_default_hyperparam = optimizer.Hyperparameter()
+_default_hyperparam.alpha = 0.001
+_default_hyperparam.beta1 = 0.9
+_default_hyperparam.beta2 = 0.999
+_default_hyperparam.beta3 = 0.999
+_default_hyperparam.c = 10.0
+_default_hyperparam.eps = 1e-8
+_default_hyperparam.eta = 1.0
+_default_hyperparam.f_star = 0.0
+_default_hyperparam.weight_decay_rate = 0
+_default_hyperparam.amsgrad = False
+
+
+def _learning_rate(hp, t, d_tilde):
+    if t == 0:
+        raise RuntimeError(
+            'Can\'t determine the learning rate of Eve optimizer '
+            'because the update steps have not been started.')
+    fix1 = 1. - math.pow(hp.beta1, t)
+    fix2 = 1. - math.pow(hp.beta2, t)
+    return (hp.alpha / d_tilde) * math.sqrt(fix2) / fix1
+
+
+class EveRule(adam.AdamRule):
+
+    """Update rule of Eve optimization algorithm.
+
+    See: https://arxiv.org/abs/1611.01505v3
+
+    Before calling :meth:`update`, :attr:`d_tilde` must be set.
+
+    Args:
+        parent_hyperparam (~chainer.optimizer.Hyperparameter): Hyperparameter
+            that provides the default values.
+        alpha (float): Coefficient of learning rate.
+        beta1 (float): Exponential decay rate of the first order moment.
+        beta2 (float): Exponential decay rate of the second order moment.
+        eps (float): Small value for the numerical stability.
+        eta (float): Schedule multiplier, can be used for warm restarts.
+        weight_decay_rate (float): Weight decay rate.
+        amsgrad (bool): Whether to use the AMSGrad variant of Eve.
+    """
+
+    d_tilde = None
+
+    @property
+    def lr(self):
+        assert self.d_tilde is not None
+        return _learning_rate(self.hyperparam, self.t, self.d_tilde)
+
+
+class Eve(optimizer.GradientMethod):
+
+    """Eve optimizer.
+
+    See: https://arxiv.org/abs/1611.01505v3
+
```

Comment:
On `weight_decay_rate` and `amsgrad`, could you refer the documentation of `Adam`?